include/deal.II/numerics/data_out_dof_data.templates.h

```diff
@@ -1170,7 +1173,7 @@ DataOut_DoFData<DoFHandlerType, patch_dim, patch_space_dim>::
       case type_dof_data:
         Assert(dof_handler != nullptr,
                Exceptions::DataOutImplementation::ExcNoDoFHandlerSelected());
-        Assert(data_vector.size() == dof_handler->n_dofs(),
+        Assert(data_vector.size() == dof_handler->n_dofs(mg_level),
```

Comment:
Without having looked into all mechanisms in detail yet, I do not believe this works in case `mg_level` is in the default `-1` state, but you have to switch at the level of the call info `n_dofs()` or `n_dofs(mg_level)`.